### PR TITLE
feat: Sync latest blocks across clients

### PR DESCRIPTION
**Pull request by nicholaspai**
# Motivation

We see the following situation which leads to inefficient use of relayer capital
- Relayer runs every 2 mins. Its likely that relayer runs overlap and multiple instances are running simultaneously.
- There is a pending deposit on chain X that is very large and requires the relayer to send additional funds to X to fill it.
- Relayer chooses not to zero fill the deposit because it will have enough funds post-rebalance to fast fill the deposit. Run 1 sends a inventory rebalance txn to send tokens to chain X.
- Relayer 2's [`SpokePoolClient.update`](https://github.com/across-protocol/relayer-v2/blob/master/src/relayer/RelayerClientHelper.ts#L140) starts before run 1's inventory rebalance txn is sent so it doesn't see the inventory rebalance event. This update can take a long time. Relayer 2 then updates the TokenClient, which does see that its balance has decreased on mainnet due to the inventory rebalance. Relayer 2 then sees the deposit but doesn't see the cross-chain rebalance AND doesn't think it has enough balance anymore to move its inventory to X, so it sends the zero fill.
- The zero fill then forces the relayer on a subsequent run to fast fill the deposit with its inventory rebalance and take repayment on the destination chain X as opposed to mainnet.

# Implementation
Syncs the `latest` block queried for the following clients:
- SpokePoolClient
- CrossChainTransferClient
- HubPoolClient
- ConfigStoreClient

This way the events queried are queried up to the same end block per chain.

Additional queries to `ERC20.balanceOf` and `provider.getBalance` to query token and ETH balances are still made with an implied `latest` block tag in the following cases:
- checking ETH balance when deciding whether to unwrap ETH in the `inventoryClient.unwrapWeth()` function
- Checking that ERC20 balances haven't changed from `TokenClient.getBalance` before executing rebalances in `inventoryClient.rebalanceInventory`
- Dataworker checking ETH balance before executing pool rebalance and relayer refund leaves

Fixes #ACX-1716

### CODE REVIEW

File: src/clients/TokenClient.ts [hunk 1]
```diff
@@ -42,13 +42,6 @@ export class TokenClient {
     return this.tokenData[chainId][token].balance;
   }
 
-  getAllowanceOnChain(chainId: number, token: string): BigNumber {
```
Comment (nicholaspai) [inline src/clients/TokenClient.ts:45]: Opportunistically removing this function that is not used anywhere. This limits the error surface involved when reading `allowance` from this client that is possible out of date

File: src/relayer/RelayerClientHelper.ts [hunk 2]
```diff
@@ -127,8 +127,6 @@ export async function constructRelayerClients(
 export async function updateRelayerClients(clients: RelayerClients, config: RelayerConfig): Promise<void> {
   // SpokePoolClient client requires up to date HubPoolClient and ConfigStore client.
   const { configStoreClient, spokePoolClients, ubaClient } = clients;
-
-  await configStoreClient.update();
```
Comment (pxrl) [inline src/relayer/RelayerClientHelper.ts:131]: Is this already handled somewhere else?

File: src/clients/TokenClient.ts [hunk 3]
```diff
@@ -219,11 +212,14 @@ export class TokenClient {
       .getAllOriginTokens()
       .map((address) => new Contract(address, ERC20.abi, spokePoolClient.spokePool.signer));
 
+    const blockTag = spokePoolClient.eventSearchConfig.toBlock ?? "latest";
```
[the remark this answers is not in the document]
Comment (pxrl) [reply src/clients/TokenClient.ts:215]: My bad - I missed that you'd addressed this in the PR description.

File: src/common/ClientHelper.ts [hunk 4]
```diff
@@ -142,14 +142,6 @@ export async function constructSpokePoolClientsWithStartBlocks(
     );
   }
 
-  logger.debug({
-    at: "ClientHelper#constructSpokePoolClientsWithStartBlocks",
-    message: "Enabled chains in block range",
-    startBlocks,
-    toBlockOverride,
-    enabledChains,
-  });
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply src/common/ClientHelper.ts:151]: We can retain it I thought it was just noisy

File: src/common/ClientHelper.ts [hunk 5]
```diff
@@ -165,7 +157,24 @@ export async function constructSpokePoolClientsWithStartBlocks(
     })
   );
 
-  return getSpokePoolClientsForContract(logger, hubPoolClient, config, spokePools, startBlocks, toBlockOverride);
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply src/common/ClientHelper.ts:168]: I think we might want to remove the `toBlockOverride` as we never use it right?

File: src/relayer/RelayerClientHelper.ts (hunk 2 above)
Comment (nicholaspai) [reply src/relayer/RelayerClientHelper.ts:131]: yeah in  `ClientHelper.updateClients` which gets called by `RelayerClientHelper.constructRelayerClients`

File: src/common/ClientHelper.ts [hunk 6]
```diff
@@ -181,8 +202,15 @@ export function getSpokePoolClientsForContract(
   config: CommonConfig,
   spokePools: { chainId: number; contract: Contract; registrationBlock: number }[],
   fromBlocks: { [chainId: number]: number },
-  toBlocks: { [chainId: number]: number } = {}
+  toBlocks: { [chainId: number]: number }
```
Comment (nicholaspai) [inline src/common/ClientHelper.ts:205]: This is a safety feature: force caller to supply toBlocks to make this function easier to reason about

File: src/common/ClientHelper.ts [hunk 7]
```diff
@@ -222,11 +256,13 @@ export async function constructClients(
...
 ): Promise<Clients> {
-  const hubSigner = baseSigner.connect(await getProvider(config.hubPoolChainId, logger));
+  const hubPoolProvider = await getProvider(config.hubPoolChainId, logger);
+  const hubSigner = baseSigner.connect(hubPoolProvider);
+  const latestMainnetBlock = await hubPoolProvider.getBlockNumber();
 
   const rateModelClientSearchSettings = {
     fromBlock: Number(getDeploymentBlockNumber("AcrossConfigStore", config.hubPoolChainId)),
-    toBlock: undefined,
+    toBlock: config.toBlockOverride[config.hubPoolChainId] ?? latestMainnetBlock,
```
Comment (nicholaspai) [inline src/common/ClientHelper.ts:265]: This was arguably a bug. HubPoolClient/ConfigStoreClient should also be allowing caller to override to blocks the same way that the SpokePoolClients can be overridden

File: src/common/ClientHelper.ts [hunk 8]
```diff
@@ -192,9 +220,15 @@ export function getSpokePoolClientsForContract(
         registrationBlock,
       });
     }
+    if (!isDefined(toBlocks[chainId])) {
```
Comment (mrice32) [inline src/common/ClientHelper.ts:223]: +1!

File: src/common/ClientHelper.ts (hunk 6 above)
Comment (mrice32) [reply src/common/ClientHelper.ts:205]: +1 -- this PR removes a lot of unhappy paths that make everything harder to analyze. These small tweaks help the type enforce more guarantees.

File: src/common/ClientHelper.ts [hunk 9]
```diff
@@ -165,7 +165,28 @@ export async function constructSpokePoolClientsWithStartBlocks(
     })
   );
 
-  return getSpokePoolClientsForContract(logger, hubPoolClient, config, spokePools, startBlocks, toBlockOverride);
+  // Explicitly set toBlocks for all chains so we can re-use them in other clients to make sure they all query
+  // state to the same "latest" block per chain.
+  const latestBlocksForChain: Record<number, number> = Object.fromEntries(
```
Comment (mrice32) [inline src/common/ClientHelper.ts:170]: This could be a dumb question, but why do this here instead of doing it at a higher level where we can pass the results _both_ to the spoke clients, the hub client, and the other clients?

It just seems odd that we're reading from the hubPoolClient to inform the values, when it seems like this function (if performed earlier) could _supply_ the info to the HubPoolClient constructor instead of reading from it.

File: src/clients/bridges/BaseAdapter.ts [hunk 10]
```diff
@@ -96,17 +95,11 @@ export abstract class BaseAdapter {
     return {
       l1SearchConfig: {
         ...this.baseL1SearchConfig,
-        fromBlock: this.baseL1SearchConfig.toBlock
-          ? this.baseL1SearchConfig.toBlock + 1
-          : this.baseL1SearchConfig.fromBlock,
-        toBlock: l1LatestBlock,
```
Comment (mrice32) [inline src/clients/bridges/BaseAdapter.ts:102]: Why was this removed? I don't see where this functionality is replicated elsewhere. Was it no longer needed?

File: src/clients/TokenClient.ts (hunk 1 above)
Comment (mrice32) [reply src/clients/TokenClient.ts:45]: +1 removing dead code is always good!

**Review by mrice32 — COMMENTED** [review]
Looks great! I think this contains a lot of good tweaks, reducing the number of paths the code can flow through. A few minor questions!

File: src/clients/bridges/BaseAdapter.ts (hunk 10 above)
Comment (nicholaspai) [reply src/clients/bridges/BaseAdapter.ts:102]: This function is supposed to return a `from/toBlock` for L1 and L2. Currently, it sets `fromBlock` = `this.baseL1SearchConfig.fromBlock` because `this.baseL1SearchConfig.toBlock` is pretty much always `undefined`. This is a weird conditional  that I wanted to simplify so I believe its equivalent to just return the `baseL1SearchConfig`.

In total, I think I'm duplicating the logic but making it more readable and also making sure that the `toBlock` returned is equal to the eventSearchConfig.toBlock of the spoke pool clients

File: src/common/ClientHelper.ts (hunk 9 above)
Comment (nicholaspai) [reply src/common/ClientHelper.ts:170]: The main reason is that we need to first figure out which chains we're going to create `spokeSigners` for, and then load latest blocks for all chains. I think we can refactor this though pretty easily I'll take a crack at it

Comment (nicholaspai) [reply src/common/ClientHelper.ts:170]: Ah yes, ok so this is an optimization. In order to figure out which chains we need to create spoke pool signers for, to subsequently query `latest` blocks, we dynamically try to load the enabled chains from the config store client. This requires the config store client to be updated so the pattern in all the bots is to call `constructClients` to create the config store client (amongst other clients), update it, then create spoke pool clients.

We could change this, but it would require keeping a hard coded listing somewhere of all chains we enable rather than being able to load them dynamically from the config store client. I think we actually used to maintain a hardcoded listing (i.e. a `CHAIN_ID_INDICES` default list) but we found this tedious to maintain so we removed it.

File: src/common/ClientHelper.ts [hunk 11]
```diff
@@ -239,10 +275,8 @@ export async function constructClients(
   );
 
   const hubPoolClientSearchSettings = {
+    ...rateModelClientSearchSettings,
     fromBlock: Number(getDeploymentBlockNumber("HubPool", config.hubPoolChainId)),
-    toBlock: undefined, // Important that we set this to `undefined` to always look up latest HubPool events such as
-    // ProposeRootBundle in order to match a bundle block evaluation block range with a pending root bundle.
```
[the remark this answers is not in the document]
Comment (pxrl) [reply src/common/ClientHelper.ts:249]: Should we not set `toBlock: latestMainnetBlock` instead ?

[the remark this answers is not in the document]
Comment (nicholaspai) [reply src/common/ClientHelper.ts:249]: We do, we actually inherit it from `... rateModelClientSearchSettings`
